### PR TITLE
scimap/spatiallda

### LINKED ISSUE #5166: scimap/spatiallda
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by migueLib**
This PR would close issue #5166 , Lint and test passed 

## PR checklist

Closes #5166 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 1]
```diff
@@ -0,0 +1,36 @@
+process SCIMAP_SPATIALLDA {
+    tag "$meta.id"
+    label 'process_single'
+
+    container "docker.io/miguelib/scimap:0.0.3"
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/main.nf:5]: Use proper container declaration, see other modules for reference. Quay (quay.io) is normally used for containers, push the container there as well if possible.

File: modules/nf-core/scimap/spatiallda/environment.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+name: "scimap_spatiallda"
+channels:
+  - conda-forge
+  - bioconda
+  - defaults
+dependencies:
+  - "YOUR-TOOL-HERE"
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/environment.yml:9]: If possible, add a reference to a Conda container for the tool.

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    input:
+    tuple val(meta), path(phenotyped)
+
+    output:
+    tuple val(meta), path("*.csv"), emit: spatial_lda_output
+    tuple val(meta), path("*.png"), emit: composition_plot
+    tuple val(meta), path("*.html"), emit: motif_location_plot
+    path "versions.yml"           , emit: versions
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/main.nf:14]: Align commas

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/main.nf:21]: Align equal signs

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        --output "${prefix}.csv" \
+        --neighborhood-composition-plot "${prefix}.png" \
+        --motif-locations-plot "${prefix}.html" \
+        $args
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        scimap/spatialLDA: \$(python /scimap/scripts/spatialLDA.py --version)
+    END_VERSIONS
+    """
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/main.nf:35]: I see this module performs some computation, please add a stub script for it

File: modules/nf-core/scimap/spatiallda/tests/main.nf.test [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                    file(params.test_data['imaging']['downstream']['phenotyped'], checkIfExists: true)
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success }
+            )
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/tests/main.nf.test:28]: Consider adding assertions for output files, preferably through snapshot if they are deterministic, or through existence/header check if not. Check the documentation and feel free to ask for details.

File: modules/nf-core/scimap/spatiallda/tests/main.nf.test [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    }
```
Comment (itrujnara) [inline modules/nf-core/scimap/spatiallda/tests/main.nf.test:30]: Consider adding a stub test, see other modules for reference

**Review by itrujnara — COMMENTED** [review]
Solid work. There is a need for some minor improvements and additions, check the code comments.

File: modules/nf-core/scimap/spatiallda/environment.yml (hunk 2 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/environment.yml:9]: Unfortunately, there's no conda container for this tool, should we delete the bioconda line together with dependencies?

Comment (itrujnara) [reply modules/nf-core/scimap/spatiallda/environment.yml:9]: You can try, although I'm not sure how it works with CI tests

File: modules/nf-core/scimap/spatiallda/tests/main.nf.test (hunk 6 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/tests/main.nf.test:52]: We tried but it's non deterministic

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 5 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf:35]: Added

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 4 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf:21]: Done

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 3 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf]: Done

File: modules/nf-core/scimap/spatiallda/environment.yml (hunk 2 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/environment.yml]: We deleted the environment.yml

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 1 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf:5]: As it's not in bioconda and there are no current plans of adding it to it we can't add it to quay.io

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 8]
```diff
@@ -0,0 +1,50 @@
+process SCIMAP_SPATIALLDA {
+    tag "$meta.id"
+    label 'process_single'
+
+    container "docker.io/miguelib/scimap:0.0.3"
```
Comment (kbestak) [inline modules/nf-core/scimap/spatiallda/main.nf:5]: I would suggest creating a repository with the Dockerfile, continuous integration and documentation about required and optional CLI parameters for clearer usage in MCMICRO (or other pipelines). Additionally, the build should be done to support both arm64 and amd64 architectures if possible.

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    tuple val(meta), path("*.png") , emit: composition_plot
+    tuple val(meta), path("*.html"), emit: motif_location_plot
+    path "versions.yml"            , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args   = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (kbestak) [inline modules/nf-core/scimap/spatiallda/main.nf:21]: Align ?: as well please

File: modules/nf-core/scimap/spatiallda/main.nf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    """
+    touch "${prefix}.csv"
+    touch "${prefix}.png"
+    touch "${prefix}.html"
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        scimap/spatialLDA: \$(python /scimap/scripts/spatialLDA.py --version)
```
Comment (kbestak) [inline modules/nf-core/scimap/spatiallda/main.nf:47]: Which version does this print? The official scimap version you're using or your script's version?

File: modules/nf-core/scimap/spatiallda/meta.yml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      licence: ["MIT licence"]
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
+  - phenotyped:
+      type: file
+      description: Phenotyped CSV file
```
Comment (kbestak) [inline modules/nf-core/scimap/spatiallda/meta.yml:23]: Provide clearer description of the csv - e.g. recommended/required columns here

File: modules/nf-core/scimap/spatiallda/tests/main.nf.test.snap
Comment (kbestak) [inline modules/nf-core/scimap/spatiallda/tests/main.nf.test.snap:1]: The md5sums for the outputs are the same (d41d8cd98f00b204e9800998ecf8427e). Please check the emitted outputs.

**Review by kbestak — CHANGES_REQUESTED** [review]
* suggestion: ghcr.io hosted image with builds for both amd64 and arm64 with visible Dockerfile and documentation on usage.
* minor: formatting inconsistency
* major: outputs in snap file are empty

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 9 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf:21]: needed? it passes pretier

File: modules/nf-core/scimap/spatiallda/main.nf (hunk 10 above)
Comment (migueLib) [reply modules/nf-core/scimap/spatiallda/main.nf:47]: Scripts version which should match the scimap version

**Review by kbestak — APPROVED** [review]
Thanks for addressing the issues, LGTM!

**Pull request merged**